tests/core/test_flow.py

```diff
@@ -3252,3 +3254,70 @@ def test_slugs_robust_to_task_name_changes(self):
             flow.add_task(a2)
 
         assert flow.slugs == {a1: "a-1", a2: "a-2"}
+
+
+class TestTerminalStateHandler:
+    def test_terminal_state_handler_determines_final_state(self):
+        def fake_terminal_state_handler(
+            flow: Flow,
+            state: State,
+            task_states: Dict[Task, State],
+        ) -> Optional[State]:
+            task_i_really_care_about = "fake_2"
+            for task, task_state in task_states.items():
+                if task.name == task_i_really_care_about and task_state.is_successful():
+                    state.message = "Custom message here"
+            return state
+
+        with Flow("test", terminal_state_handler=fake_terminal_state_handler) as flow:
+            fake_task = Task("fake")
+            flow.add_task(fake_task)
+            fake_task_2 = Task("fake_2")
+            flow.add_task(fake_task_2)
+            fake_task_2.set_upstream(fake_task)
+
+        assert flow.terminal_state_handler == fake_terminal_state_handler
+        assert flow.run().message == "Custom message here"
+
+    def test_flow_state_used_if_terminal_state_handler_does_not_return_a_new_state(
+        self,
+    ):
+        def fake_terminal_state_handler(
+            flow: Flow,
+            state: State,
+            task_states: Dict[Task, State],
+        ) -> Optional[State]:
+            return None
+
+        with Flow("test", terminal_state_handler=fake_terminal_state_handler) as flow:
+            fake_task = Task("fake")
+            flow.add_task(fake_task)
+
+        assert flow.run().message == "All reference tasks succeeded."
```

Comment:
I have a minor preference for asserting the final state type is still `Success` as well instead of just the message